rfc/0000-batch-cancellation.rst

```diff
@@ -0,0 +1,538 @@
+until we reach the number of cores the next user has currently
+running. We then divide up the remaining cores equally amongst the two
+users until we reach the number of cores the next user has running. We
+repeat until we have either exhausted all free cores in the cluster or
+have satisfied all user resource requests. These fair share
+allocations are then translated into a concrete number of jobs that
+can be scheduled per user in each autoscaling loop. We then find the
+sum of CPU resources requested for each user's share of jobs and then
+apply some further optimizations to account for which regions a job
+can run in to come up with the final number of worker VMs that should
+be provisioned. The query to get the number of ready cores in the fair
+share algorithm is fast because we aggregate across a global table
+``user_inst_coll_resources`` that has a limited number of rows
+maintaining counts of the number of ready cores per instance
+collection and user.
+
+The key insight in our current system for a fast cancellation
+implementation is to be able to do an O(1) update of the
+``user_inst_coll_resources`` table that keeps track of the number of
+ready cores and jobs upon cancelling the batch by subtracting the
+number of **cancellable** resources and jobs for that batch. This way
+both the autoscaler and scheduler instantly see this change in demand
+with their respective fair share algorithms. In addition, the
+autoscaler will not provision unnecessary resources for the cancelled
+batch as the query that scans the head of the jobs table will ignore
+jobs that are "Ready" but in batches that have been
+cancelled. Likewise, the SQL query the scheduler uses will ignore
+"Ready" jobs for batches that have been cancelled and the fair share
+algorithm will instantly adapt to the change in demand from the
+cancelled batch to schedule other users and batches fairly.
+
+More concretely, there are two tables that keep track of resources:
+
+- ``user_inst_coll_resources``: Keeps track of the total number of
+  jobs and cores for each user and instance collection for jobs in
+  Ready, Running, and Creating states as well as the number of
+  cancelled jobs in each state. Note this table is not parameterized
+  by batch ID.
+- ``batch_inst_coll_cancellable_resources``: Keeps track of the total
+  number of **cancellable** jobs and cores for jobs in Ready, Running,
+  and Creating states. Note this table is parameterized by
+  batch id. The numbers in this table do not include counts for
+  **always_run** jobs as they cannot be cancelled.
+
+These two tables are tokenized for fast concurrent
+updates. Tokenization means that the value for a single key in the
+database is represented by up to 200 rows in order to avoid
+serialization of updates. Therefore, to compute the number of jobs or
+free cores, the query must do an aggregation for all rows for a given
+key which can be up to 200 rows.
```

Comment:
I think the discussion of tokenization is an implementation detail that is unimportant for this discussion. It is sufficient and correct to say that querying these tables is O(1).